Explorer/Assets/DCL/Multiplayer/Connections/Messaging/MessageWrap.cs

```diff
@@ -33,27 +37,68 @@ public MessageWrap(T payload, IDataPipe dataPipe, List<string> recipients, IMult
             sent = false;
         }
 
-        public void Send(DataPacketKind dataPacketKind = DataPacketKind.KindLossy)
+        public async UniTaskVoid SendAndDisposeAsync(DataPacketKind dataPacketKind = DataPacketKind.KindLossy)
         {
             if (sent)
                 throw new Exception("Request already sent");
 
-            using MemoryWrap memory = memoryPool.Memory(Payload);
-            Payload.WriteTo(memory);
+            await UniTask.SwitchToThreadPool();
+            using var packetWrap = multiPool.TempResource<Packet>();
+            packetWrap.value.ClearMessage();
+            WritePayloadToPacket(packetWrap.value);
+            using MemoryWrap memory = memoryPool.Memory(packetWrap.value);
+            packetWrap.value.WriteTo(memory);
             dataPipe.PublishData(memory.Span(), TOPIC, recipients, dataPacketKind);
             sent = true;
+            Dispose();
         }
 
         public void AddRecipient(string sid)
         {
             recipients.Add(sid);
         }
 
-        public void Dispose()
+        private void WritePayloadToPacket(Packet packet)
+        {
+            var type = typeof(T);
+
+            if (MessageWrapExtensions.WRITES_MAP.TryGetValue(type, out var writeAction) == false)
```

Comment:
It looks like this action can be resolved and stored statically once as `T` is a generic argument of the whole structure